les/api.go

```diff
@@ -17,16 +17,407 @@
+	info := make(map[string]interface{})
+	if c != nil {
+		now := mclock.Now()
+		info["isConnected"] = true
+		info["connectionTime"] = float64(now-c.connectedAt) / float64(time.Second)
+		info["capacity"] = c.capacity
+		pb, nb := c.balanceTracker.getBalance(now)
+		info["pricing/balance"], info["pricing/negBalance"] = pb, nb
+		info["pricing/balanceMeta"] = c.balanceMetaInfo
+		info["priority"] = pb != 0
+	} else {
+		info["isConnected"] = false
+		pb := api.server.clientPool.getPosBalance(id)
+		info["pricing/balance"], info["pricing/balanceMeta"] = pb.value, pb.meta
+		info["priority"] = pb.value != 0
+	}
+	return info
+}
+
+// sendEvent sends an event to the subscribers interested in it. For global events client == nil.
+func (api *PrivateLightServerAPI) sendEvent(clientEvent string, client *clientInfo) {
+	if len(api.subs) == 0 {
+		return
+	}
+
+	event := make(map[string]interface{})
+	event["totalCapacity"], event["totalConnectedCapacity"], event["priorityConnectedCapacity"] = api.server.clientPool.capacityInfo()
+	if client != nil {
+		event["clientEvent"] = clientEvent
+		event["clientId"] = client.id
+		event["clientInfo"] = api.clientInfo(client, client.id)
+	}
+
+	for sub := range api.subs {
+		select {
+		case <-sub.rpcSub.Err():
+			delete(api.subs, sub)
+		case <-sub.notifier.Closed():
+			delete(api.subs, sub)
+		default:
+			sub.notifier.Notify(sub.rpcSub.ID, event)
+		}
+	}
+}
+
+// setParams either sets the given parameters for a single client (if ID is specified)
+// or the default parameters applicable to clients connected in the future
+func (api *PrivateLightServerAPI) setParams(params map[string]interface{}, client *clientInfo, id enode.ID, posFactors, negFactors *priceFactors) (updateFactors bool, err error) {
+	if client != nil {
+		posFactors, negFactors = &client.posFactors, &client.negFactors
+	}
+	defParams := id == enode.ID{}
+loop:
+	for name, value := range params {
+		errValue := func() error {
+			return fmt.Errorf("invalid value for parameter '%s'", name)
+		}
+		setFactor := func(v *float64) {
+			if posFactors != nil {
+				if val, ok := value.(float64); ok && val >= 0 {
+					*v = val / float64(time.Second)
+					updateFactors = true
+				} else {
+					err = errValue()
+				}
+			} else {
+				err = errClientNotConnected
+			}
+		}
+
+		processed := true
+		switch name {
+		case "pricing/timeFactor":
+			setFactor(&posFactors.timeFactor)
+		case "pricing/capacityFactor":
+			setFactor(&posFactors.capacityFactor)
+		case "pricing/requestCostFactor":
+			setFactor(&posFactors.requestFactor)
+		case "pricing/negative/timeFactor":
+			setFactor(&negFactors.timeFactor)
+		case "pricing/negative/capacityFactor":
+			setFactor(&negFactors.capacityFactor)
+		case "pricing/negative/requestCostFactor":
+			setFactor(&negFactors.requestFactor)
+		default:
+			processed = false
+			if defParams {
+				err = fmt.Errorf("invalid default parameter '%s'", name)
```

Comment:
If there is any error, why not return directly?

Reply:
This logic actually might make sense because if we are setting many parameters and returning in the middle of it because of an error then it is not clear which parameters have been changed. This way it is clear that every valid parameter change has happened.
Checking for all possible errors before changing anything would be another logical alternative but that would probably make the code unnecessarily complex.

Reply:
But the error will be overwritten over and over again.  If if we have 2 invaid parameters, it's also unclear

Reply:
Ok, I changed the error handling. Now it returns on the first error so the caller has to retry until the call returns with no error. Now that adding balance is a separate function and not handled by `setParams`, setting something twice is not a problem.